docs/rfcs/IntelGpuDialect.md

```diff
@@ -0,0 +1,122 @@
+
+
+#### intel_gpu.destroy_stream (intel_gpu::DestroyStreamOp)
+
+Destroy Intel GPU Stream.
+
+Syntax:
+
+```
+operation ::= "intel_gpu.destroy_stream"() : () -> !intel_gpu.OpaqueType
+```
+
+The intel_gpu.destroy_stream() operation will be added by a pass called AddIntelGpuStream after the dialect is in Gpu Dialect. After that the IR is lowered all the way down to LLVM.
+This operation will destroy the above created stream.
+
+
+Example Usage of lowering:
+
+This is how the IR will look after inserting the create_stream & destroy_stream operation in the IR via a custom pass called AddIntelGpuStream.
+The create_stream op will be inserted at the begining of the IR before the device allocation ops because the device allocation requires the stream as one of the input.
+
+```
+
+func.func @main() {
+    %c8 = arith.constant 8 : index
+    %c1 = arith.constant 1 : index
+    %cst = arith.constant 2.200000e+00 : f32
+    %cst_0 = arith.constant 1.100000e+00 : f32
+    %cst_1 = arith.constant 0.000000e+00 : f32
+    %0 = "intel_gpu.create_stream"() : () -> !intel_gpu.OpaqueType
+    .
+    .
+    .
+    .
+    %6 = gpu.launch_func async [%5] @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%memref : memref<8xf32>, %memref_2 : memref<8xf32>, %memref_4 : memref<8xf32>)
+    %7 = memref.cast %memref_4 : memref<8xf32> to memref<*xf32>
+    gpu.wait [%6]
+    call @printMemrefF32(%7) : (memref<*xf32>) -> ()
+    "intel_gpu.destroy_stream"(%0) : (!intel_gpu.OpaqueType) -> ()
+    return
+  }
+
+  ```
+
+After lowering to LLVM the IR will be transformed into:
+
+```
+
+func.func @main() {
+    %c8 = arith.constant 8 : index
+    %0 = builtin.unrealized_conversion_cast %c8 : index to i64
+    %c1 = arith.constant 1 : index
+    %1 = builtin.unrealized_conversion_cast %c1 : index to i64
+    %cst = arith.constant 2.200000e+00 : f32
+    %cst_0 = arith.constant 1.100000e+00 : f32
+    %cst_1 = arith.constant 0.000000e+00 : f32
+    %3 = llvm.call @iGpuCreateStream() : (!llvm.ptr<i8>) -> !llvm.ptr<i8>
+    %4 = llvm.mlir.constant(8 : index) : i64
+    %5 = llvm.mlir.constant(1 : index) : i64
+    %6 = llvm.mlir.null : !llvm.ptr<f32>
+    %7 = llvm.getelementptr %6[%4] : (!llvm.ptr<f32>, i64) -> !llvm.ptr<f32>
+    %8 = llvm.ptrtoint %7 : !llvm.ptr<f32> to i64
+    %9 = llvm.mlir.constant(64 : i64) : i64
+    %10 = llvm.call @iGpuMemAlloc(%8, %9, %3) : (i64, i64, !llvm.ptr<i8>) -> !llvm.ptr<i8>
+    .
+    .
+    .
+    %62 = llvm.call @iGpuModuleLoad(%3, %60, %61) : (!llvm.ptr<i8>, !llvm.ptr<i8>, i64) -> !llvm.ptr<i8>
+    .
+    %66 = llvm.call @iGpuKernelGet(%3, %62, %65) : (!llvm.ptr<i8>, !llvm.ptr<i8>, !llvm.ptr<i8>) -> !llvm.ptr<i8>
+    .
+    llvm.call @iGpuLaunchKernel(%3, %66, %0, %1, %1, %1, %1, %1, %67, %108, %109) : (!llvm.ptr<i8>, !llvm.ptr<i8>, i64, i64, i64, i64, i64, i64, i32, !llvm.ptr<struct<(ptr<i8>, i64)>>, !llvm.ptr<ptr<i8>>) -> ()
```

Comment:
If user creates multiple streams, how the lowering decides which stream to use?  I think user should pass the stream to launch kernel function.

Reply:
the launch kernel takes stream as an input when its lowered to llvm IR..
```llvm.call @iGpuLaunchKernel(%3, %66, %0, %1, %1, %1, %1, %1, %67, %108, %109)```

here %3 is the stream/queue

Reply:
How does the lowering decide which stream to use?

Reply:
This RFC needs to be changed. User decide which stream to use for each kernel, not compiler.